src/main/java/au/org/ala/names/index/ALANameAnalyser.java

```diff
@@ -0,0 +1,532 @@
+            if (r.getStrRanks() != null) {
+                for (String label : r.getStrRanks())
+                    this.rankMap.put(label.toUpperCase().trim(), r);
+            }
+        }
+        this.loadCsv(DEFAULT_RANK_CODE_MAP, this.rankMap, RankType.class);
+    }
+    /**
+     * Build a synonym map.
+     */
+    protected void buildNomenclaturalStatusMap() {
+        this.nomenclaturalStatusMap = new HashMap<>(64);
+        // Use NomStatusParser for default values
+        this.loadCsv(DEFAULT_NONEMCLATURAL_STATUS_CODE_MAP, this.nomenclaturalStatusMap, NomenclaturalStatus.class);
+    }
+
+    protected void buildInformalPatternList() {
+        this.informalPatterns = new ArrayList<>(32);
+        this.loadPatternCsv(DEFAULT_INFORMAL_PATTERN_LIST, this.informalPatterns);
+    }
+
+    /**
+     * Canonicalise the nomenclatural code.
+     *
+     * @param code The code name
+     *
+     * @return The mapped code or null for not found
+     */
+    @Override
+    public NomenclaturalCode canonicaliseCode(String code) {
+        if (code == null)
+            return null;
+        code = code.toUpperCase().trim();
+        NomenclaturalCode nc = this.codeMap.get(code);
+        if (nc == null)
+            this.report(IssueType.PROBLEM, "nomenclaturalCode.notFound", code);
+        return nc;
+    }
+
+    /**
+     * Canonicalise the taxonomic status.
+     * <p>
+     * If the term cannot be parsed, {@link TaxonomicType#INFERRED_UNPLACED} is used.
+     * </p>
+     *
+     * @param taxonomicStatus The taxonomic status term
+     *
+     * @return The mapped status
+     */
+    @Override
+    public TaxonomicType canonicaliseTaxonomicType(String taxonomicStatus) {
+        if (taxonomicStatus == null)
+            return TaxonomicType.INFERRED_UNPLACED;
+        taxonomicStatus = taxonomicStatus.toUpperCase().trim();
+        if (taxonomicStatus.isEmpty())
+            return TaxonomicType.INFERRED_UNPLACED;
+        TaxonomicType type = this.taxonomicTypeMap.get(taxonomicStatus);
+        if (type == null) {
+            this.report(IssueType.PROBLEM, "taxonomicStatus.notFound", taxonomicStatus);
+            type = TaxonomicType.INFERRED_UNPLACED;
+            synchronized (this) {
+                this.taxonomicTypeMap.put(taxonomicStatus, type); // Report once
+            }
+        }
+        return type;
+    }
+
+    /**
+     * Canonicalise the rank.
+     *
+     * @param rank The rank term
+     *
+     * @return The mapped rank
+     */
+    @Override
+    public RankType canonicaliseRank(String rank) {
+        if (rank == null)
+            return RankType.UNRANKED;
+        rank = rank.toUpperCase().trim();
+        if (rank.isEmpty())
+            return RankType.UNRANKED;
+        RankType rankType = this.rankMap.get(rank);
+        if (rankType == null) {
+            this.report(IssueType.PROBLEM, "rank.notFound", rank);
+            rankType = RankType.UNRANKED;
+            synchronized (this) {
```

Comment:
Overwriting is idempotent, since they're all basically just placeholders for unrecognised terms.

The vocabularies for this should be moved to the configuration file. However, it's not broken at the moment and needs future work.